### ISSUE TITLE
logspout fails to reconnect after connection reset by peer.

### ISSUE THREAD

**Issue by jeacott1** [gliderlabs/logspout#215]
I have 5 instances of logspout running in marathon talking to loggly.
all was well until yesterday when all 5 failed with errors like this:

```
#   ADAPTER     ADDRESS                 CONTAINERS      SOURCES OPTIONS
#   syslog+tcp  logs-01.loggly.com:514                          map[]
2016/07/09 03:39:42 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection reset by peer
2016/07/09 03:39:42 syslog: retrying tcp up to 11 times
2016/07/09 03:40:23 syslog: retry failed
2016/07/09 03:40:23 syslog: reconnecting up to 11 times
2016/07/10 23:33:50 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/10 23:33:50 syslog: retrying tcp up to 11 times
2016/07/10 23:34:31 syslog: retry failed
2016/07/10 23:34:31 syslog: reconnecting up to 11 times
2016/07/11 02:04:47 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/11 02:04:47 syslog: retrying tcp up to 11 times
2016/07/11 02:05:28 syslog: retry failed
2016/07/11 02:05:28 syslog: reconnecting up to 11 times
2016/07/11 02:32:28 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/11 02:32:28 syslog: retrying tcp up to 11 times
2016/07/11 02:33:09 syslog: retry failed
2016/07/11 02:33:09 syslog: reconnecting up to 11 times
2016/07/12 02:57:56 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/12 02:57:56 syslog: retrying tcp up to 11 times
2016/07/12 02:58:37 syslog: retry failed
2016/07/12 02:58:37 syslog: reconnecting up to 11 times
2016/07/13 23:51:10 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: broken pipe
2016/07/13 23:51:10 syslog: reconnecting up to 11 times
2016/07/18 11:11:16 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: broken pipe
2016/07/18 11:11:16 syslog: reconnecting up to 11 times
2016/07/22 05:51:14 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: broken pipe
2016/07/22 05:51:14 syslog: reconnecting up to 11 times
2016/07/27 08:00:46 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection reset by peer
2016/07/27 08:00:46 syslog: retrying tcp up to 11 times
2016/07/27 08:01:27 syslog: retry failed
2016/07/27 08:01:27 syslog: reconnecting up to 11 times
2016/07/27 17:40:52 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/27 17:40:52 syslog: retrying tcp up to 11 times
2016/07/27 17:41:33 syslog: retry failed
2016/07/27 17:41:33 syslog: reconnecting up to 11 times
2016/07/28 00:56:24 syslog: write tcp xxx.xxx.xxx.xxx:34657->logglyip:514: write: connection timed out
2016/07/28 00:56:24 syslog: retrying tcp up to 11 times
2016/07/28 00:57:05 syslog: retry failed
2016/07/28 00:57:05 syslog: reconnecting up to 11 times
```

it looks like it retried with a backoff, but never succeeded in reconnecting, even though when I manually tried connecting it worked just fine.  

restarting the dockers fixed the issue, so 2 things from this.
1. something doesn't work as expected with the retry mechanism.
2. if there was a useful healthcheck endpoint in logspout, marathon could have shot/restarted the processes.

**Comment by Halama** [comment]
Same here, my containers stopped logging to Papertrail:
```
2016/08/04 21:01:25 syslog: write tcp 172.17.0.2:33158->67.214.223.210:49730: write: connection reset by peer
2016/08/04 21:01:25 syslog: retrying tcp up to 11 times
2016/08/04 21:02:06 syslog: retry failed
2016/08/04 21:02:06 syslog: reconnecting up to 11 times
```

It started log again after manual restart of container.

**Comment by albertsun** [comment]
I'm seeing the same thing trying to write to Papertrail. Manual restart of the service fixes things.

```
Aug 03 14:20:17 ip-10-214-0-120.ec2.internal docker[30862]: 2016/08/03 14:20:17 syslog: write tcp 192.168.16.2:44954->173.247.107.217:49069: write: connection reset by peer
Aug 03 14:20:17 ip-10-214-0-120.ec2.internal docker[30862]: 2016/08/03 14:20:17 syslog: retrying tcp up to 11 times
Aug 03 14:20:58 ip-10-214-0-120.ec2.internal docker[30862]: 2016/08/03 14:20:58 syslog: retry failed
Aug 03 14:20:58 ip-10-214-0-120.ec2.internal docker[30862]: 2016/08/03 14:20:58 syslog: reconnecting up to 11 times
```

**Comment by jacobdr** [comment]
I see the same behavior. And when I restart it looks like all of the logs have been cached, and upon restart they all get flushed out again. 

I think there is a bug in the retry logic, but the quickest way out might be to just crash the container upon timeout/connection loss and let the restart semantics of docker take care of reconnection.

**Comment by edude03** [comment]
I'm having the same issue. Anyone interested in a bounty?

**Comment by ryansch** [comment]
Honestly after hitting this bug I spent a little time on getting the docker syslog driver working and I've been pleased with the result.

**Comment by yeasy** [comment]
Any solution on the problem?
Maybe we should enable reconnection?
Thanks!

**Comment by adrien-candiotti** [comment]
Same here, i'm using dockercloud so modifying the docker syslog driver is not a good option for me.
Any idea?

**Comment by izakp** [comment]
+1 would really love this fixed as it makes logspout unusable - it never gets past the message "reconnecting up to 11 times" which makes me think it's stuck in the reconnection loop and never exiting

**Comment by andrewgaul** [comment]
I suspect that the exponential retry logic cannot handle long-lived network failures.  It retries 11 times with a maximum interval of 10240 milliseconds.

**Comment by edude03** [comment]
I think want we want is for the container to die when it can't reconnect (ideally buffered but let's not get ahead of ourselves) that way the scheduler can restart the container and/or notify someone when it's not working

**Comment by izakp** [comment]
I'd be happy if it tried 1 reconnection and on error, exited, then leaving it up to the scheduler to take further action.

**Comment by yeasy** [comment]
One possible way is to support setting retry times by configuration.

**Comment by rhuddleston** [comment]
After investigating this more I can confirm I'm seeing the same thing where after "reconnecting up to 11 times" I'm losing messages but the interesting thing is It only stops logging some of the docker containers. Others continue logging and the ones that stop logging I can fix by restarting logspout

**Comment by imriz** [comment]
for the benefit of everyone affected by this (huge) bug, and due to the lack of activity in this issue - I've forked @robcza fork, which implements both the configurable retry, and the exit on failure,
and created a public image in docker hub: https://hub.docker.com/r/imriz/logspout/

**Comment by qiang0723** [comment]
@imriz 
I got some errors when using your image  in the hub(https://hub.docker.com/r/imriz/logspout/) for testing. 
here is my steps:
1.stop logstash,   logspout get errors like this:
```
2016/12/15 11:03:51 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
2016/12/15 11:03:53 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
2016/12/15 11:03:54 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
2016/12/15 11:03:54 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
```
2.restart logstash, logspout get same errors:

```
2016/12/15 11:04:15 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
2016/12/15 11:04:16 syslog: write udp 172.17.0.2:57539->192.168.7.60:5000: write: connection refused
```
command:
```
$sudo docker  run    --rm     \
                                 -v /var/run/docker.sock:/var/run/docker.sock \   
                                 imriz/logspout  \  
                                 syslog+udp://logstash-host:5000
```
Is my setup wrong, or this image doesn't support?

**Comment by imriz** [comment]
@qiang0723 I would try to sniff the traffic with tcpdump. I suspect you are getting an ICMP unreachable packet or similar.

**Comment by imriz** [comment]
I did some more digging, and I found the following items that at least in my environment caused logspout to stop logging. This may or may not apply in your environment. We are working with Docker 1.8 (year old - apparently that is *very* old for docker ;) ), so some of the points here, mostly regarding the JSON log driver might not be true in later versions.

1.  In https://github.com/gliderlabs/logspout/blob/master/router/pump.go#L351 there is a 1 second timeout for the send method. This means that if the reconnect takes more than one second, *the log stream will be closed* and from what I witnessed, it will stop the logs from that unlucky container.

2. I observed a case where a container had a large log file (2GB - 4 days worth of logs), and when the `tail` option was set to `all`, and a `since` value was given, the docker daemon choked. I waited for the stream to start, and gave up after 15 minutes. Changing the initial `tail` from `all` to `0` solved this for me. On a stream disconnect (inactivity or other), I set the `tail` to `2000`. Do note that it means that whenever logspout start it won't digest any backlog, even if that option is set. It also means that on any reconnection of the docker stream (*docker* stream, not syslog), it will only look 2000 events back.

3. Similar to previous issue, I encountered an issue with `tail` set to a large value (I tried 6000), the docker daemon might fail parsing the JSON logs. See https://github.com/docker/docker/issues/29511 - I was told to try with newer Docker version, but that is not possible for me ATM.


Anyway, I fixed all these issue at my fork, and it is present in my docker image. I am running for about 24 hours without issues (before I had 3-4 incidents per day). If you use it, YMMV and please understand the consequences of my changes.

**Comment by robcza** [comment]
@imriz I gave up fast and I am killing the container as soon as any problem occurs. It works for my use case, however it is too aggressive approach in general. 
Your investigation is much appreciated, will you go for pull requests with your results? It would help a lot of users.

**Comment by imriz** [comment]
Sure. I would appreciate it if you could test my changes too.

**Comment by michaelshobbs** [comment]
I just merged #240 that allows you to configure the number of retries. I'm still pondering the indefinite retry method as proposed by #253

**Comment by michaelshobbs** [comment]
Potential fix + tests were merged in f18e503780b8f3abb3771ffaa97e5d2971199f9b
